### PR TITLE
fix: make copying sandbox files writeable

### PR DESCRIPTION
**Pull request by rgrinberg**
Currently, the `copy` sandbox just uses whatever permissions are set in the workspace local cache.

With this PR, it will now make all files and directories writeable.

### CODE REVIEW

File: src/dune_engine/sandbox.ml [hunk 1]
```diff
@@ -41,26 +41,39 @@ type t =
...
+             | S_DIR -> `Dir (Path.Untracked.lstat_exn src).st_perm
+             | S_REG -> `File
+             | _ -> `Other kind
+           in
+           loop what ~src ~dst:(Path.relative dst name)))
+    | `Other kind ->
+      User_error.raise
+        ~hints:
+          [ Pp.text
+              "re-run dune to delete the stale artifact, or manually delete this file"
```
Comment (snowleopard) [inline src/dune_engine/sandbox.ml:68]:
```suggestion
              "Re-run Dune to delete the stale artifact, or manually delete this file."
```

File: src/dune_engine/sandbox.ml [hunk 2]
```diff
@@ -41,26 +41,39 @@ type t =
...
+           in
+           loop what ~src ~dst:(Path.relative dst name)))
+    | `Other kind ->
+      User_error.raise
+        ~hints:
+          [ Pp.text
+              "Re-run Dune to delete the stale artifact, or manually delete this file."
+          ]
+        [ Pp.textf
+            "Failed to copy file %s of kind %s while creating a copy sandbox"
```
Comment (snowleopard) [inline src/dune_engine/sandbox.ml:71]:
```suggestion
            "Failed to copy file %s of kind %S while creating a copy sandbox"
```

File: src/dune_engine/sandbox.ml [hunk 3]
```diff
@@ -90,15 +103,22 @@ let link_function ~(mode : Sandbox_mode.some) =
...
+         in
+         copy_recursively what ~src ~dst
      | Hardlink ->
        (match Sys.win32 with
         | true -> win32_error mode
         | false -> fun src dst -> Io.portable_hardlink ~src ~dst)
      | Patch_back_source_tree ->
        (* We need to let the action modify its dependencies, so we copy
           dependencies and make them writable. *)
+       (* CR-rgrinberg: this doesn't work with directory targets *)
```
Comment (snowleopard) [inline src/dune_engine/sandbox.ml:121]:
```suggestion
       (* CR-someday rgrinberg: this doesn't work with directory targets *)
```

File: test/blackbox-tests/test-cases/directory-targets/copy-sandboxing.t [hunk 4]
```diff
@@ -60,13 +60,13 @@ mode:
...
   755
   permissions of output/y
-  444
+  644
   permissions of output/x
-  444
+  644
   permissions of output/subdir
-  755
+  754
```
Comment (snowleopard) [inline test/blackbox-tests/test-cases/directory-targets/copy-sandboxing.t:67]: I thought this PR only adds permissions. Here we are losing some?

File: src/dune_engine/sandbox.ml [hunk 5]
```diff
@@ -90,15 +103,22 @@ let link_function ~(mode : Sandbox_mode.some) =
         | false -> fun src dst -> Io.portable_symlink ~src ~dst)
      | Copy ->
        fun src dst ->
-         let { Unix.st_kind; _ } = Path.Untracked.stat_exn src in
-         copy_recursively st_kind ~src ~dst
+         let what =
+           let { Unix.st_kind; st_perm; _ } = Path.Untracked.stat_exn src in
+           match st_kind with
+           | S_DIR -> `Dir st_perm
```
Comment (snowleopard) [inline src/dune_engine/sandbox.ml:109]: Let's create a regular type for this variant and put it in a module with something like `of_unix_stats` to avoid duplicating this code.

File: src/dune_engine/sandbox.ml [hunk 6]
```diff
@@ -90,15 +103,22 @@ let link_function ~(mode : Sandbox_mode.some) =
...
+         in
+         copy_recursively what ~src ~dst
      | Hardlink ->
        (match Sys.win32 with
         | true -> win32_error mode
         | false -> fun src dst -> Io.portable_hardlink ~src ~dst)
      | Patch_back_source_tree ->
        (* We need to let the action modify its dependencies, so we copy
           dependencies and make them writable. *)
+       (* CR-someday rgrinberg: this doesn't work with directory targets *)
```
Comment (snowleopard) [inline src/dune_engine/sandbox.ml:121]: I'm confused: doesn't `copy_recursively` exist solely for supporting directory targets (or, more accurately, for supporting dependencies on them)? It seems odd to go through all the trouble and not support directory targets in the end.

**Review by snowleopard — COMMENTED** [review]
Thanks! Generally looks good but I've added some suggestions/questions.

File: src/dune_engine/sandbox.ml (hunk 6 above)
Comment (rgrinberg) [reply src/dune_engine/sandbox.ml:121]: I think @emillon just forgot to add full support for them here. I have upcoming PR's that will do so though. This PR is just focusing on the permissions.

File: src/dune_engine/sandbox.ml (hunk 5 above)
Comment (rgrinberg) [reply src/dune_engine/sandbox.ml:109]: I could do a regular variant, but it actually won't prevent duplicating the code. That's because:

1. At this call site, we get the file kind from `Unix.stat`
2. In the other call sites, we get the file kind from `Path.readdir_with_kinds`

Would you still like the regular variant?

File: test/blackbox-tests/test-cases/directory-targets/copy-sandboxing.t (hunk 4 above)
Comment (rgrinberg) [reply test/blackbox-tests/test-cases/directory-targets/copy-sandboxing.t:67]: It's because we are now copying the permissions of directories correctly. Previously, we'd just do `Path.mkdir`, now we make sure we create the directory with the correct permissions.

File: src/dune_engine/sandbox.ml [hunk 7]
```diff
@@ -41,26 +41,39 @@ type t =
...
+  let chmod = Path.Permissions.add Path.Permissions.write in
+  let rec loop what ~src ~dst =
+    match what with
+    | `File -> Io.copy_file ~chmod ~src ~dst ()
+    | `Dir perms ->
+      (match Path.Untracked.readdir_unsorted_with_kinds src with
+       | Error e -> Unix_error.Detailed.raise e
+       | Ok contents ->
+         let perms = chmod perms in
+         Path.mkdir_p ~perms dst;
```
Comment (rgrinberg) [inline src/dune_engine/sandbox.ml:54]: @snowleopard this is the line that makes it appear as if we're losing permissions. But we're actually just copying them accurately. Previously, we would just do `Path.mkdir_p` which always created things as `755`.

I could move this into a separate PR if you prefer, but to me it seemed like a similar enough of a bug (incorrect permissions) to just stuff it here.

Comment (snowleopard) [reply src/dune_engine/sandbox.ml:54]: Ah, I see, thanks.

File: src/dune_engine/sandbox.ml (hunk 5 above)
Comment (snowleopard) [reply src/dune_engine/sandbox.ml:109]: Yes, I think so. Converting stats around is boring code, and it's best to keep it in a little helper module.

File: src/dune_engine/sandbox.ml (hunk 7 above)
Comment (rgrinberg) [reply src/dune_engine/sandbox.ml:54]: @snowleopard actually, should we also add executable permissions to the directory? Without executable permissions, it's not possible to write anything to the directory anyway.

Comment (snowleopard) [reply src/dune_engine/sandbox.ml:80]: Yeah, we might as well do that. I guess it doesn't matter much in practice because unreadable directories just don't show up often.

**Comment by rgrinberg** [comment]
@snowleopard addressed all the comments

**Review by snowleopard — APPROVED** [review]
Looks good, thanks!
